Date aware examples

joewallwork opened the pull request:
A short PR to make the Tohoku example date-aware. I also noticed that the North Sea example wasn't set up properly for testing: a shorter `t_end` value was set but unused.

Commit:
North Sea: actually use shorter test time

Commit:
Tohoku: make date-aware

stephankramer commented on examples/north_sea/model_config.py:
```diff
@@ -97,15 +97,15 @@ def construct_solver(spinup=False, store_station_time_series=True, **model_optio
     t_end = (end_date - start_date).total_seconds()
     if os.getenv("THETIS_REGRESSION_TEST") is not None:
-        t_end = 5 * t_export
+        end_date = datetime.datetime(2022, 1, 1, 0, 5, tzinfo=sim_tz)
```
That's 5 minutes past midnight, no? but the timestep is in hours?

jwallwork23 replied:
Good spot! Fixed in https://github.com/thetisproject/thetis/pull/325/commits/1b073425f37bb9ba501a3a3fa8a3471257ddf1ac.

stephankramer reviewed (dismissed):
Other than that looks fine. Sorry this seems to have sat there a while - not sure you were still intending to merge

Commit:
minutes->hours

Commit:
Merge branch 'master' into date-aware-examples

jwallwork23 commented:
[Merged in `master` to fix conflicts.]

joewallwork committed:
Merge branch 'main' into date-aware-examples

joewallwork committed:
Use simulation dates rather than times

joewallwork committed:
Fix start->initial

joewallwork committed:
Remove THETIS_REGRESSION_TEST end_date change

stephankramer commented on examples/tohoku_inversion/model_config.py:
```diff
@@ -117,12 +118,19 @@ def construct_solver(bathymetry_2d, elev_init, store_station_time_series=True, *
-    t_end = 2 * 3600.0
+    # Setup temporal discretisation
     u_mag = Constant(5.0)
+    default_start_date = datetime.datetime(2011, 3, 11, 14, 46, tzinfo=sim_tz)
+    default_end_date = datetime.datetime(2011, 3, 11, 16, 46, tzinfo=sim_tz)
```
```suggestion
    default_start_date = sim_tz.localize(datetime.datetime(2011, 3, 11, 14, 46))
    default_end_date = sim_tz.localize(datetime.datetime(2011, 3, 11, 16, 46))
```
If you do:
`
print(datetime.datetime(2011, 3, 11, 14, 46, tzinfo=sim_tz), sim_tz.localize(datetime.datetime(2011, 3, 11, 14, 46)))
`
you get
`
2011-03-11 14:46:00+09:19 2011-03-11 14:46:00+09:00
`
i.e. the former uses a time zone of UTC+9:19 instead of the correct UTC+9. If I understand correctly, datetime in the standard library only has limited support for timezones, and it doesn't handle the fact that timezones change historically (i.e. Japan used to be on UTC+9:19 but is now UTC+9

stephankramer commented on examples/tohoku_inversion/model_config.py:
```diff
@@ -117,12 +118,19 @@ def construct_solver(bathymetry_2d, elev_init, store_station_time_series=True, *
     if os.getenv("THETIS_REGRESSION_TEST") is not None:
-        t_end = 5 * t_export
+        model_options["simulation_initial_date"] = default_start_date
+        model_options["simulation_end_date"] = datetime.datetime(2011, 3, 11, 14, 51, tzinfo=sim_tz)
```
```suggestion
        model_options["simulation_end_date"] = sim_tz.localize(datetime.datetime(2011, 3, 11, 14, 51))
```

stephankramer committed:
Fix timedelta calculation

Affects when options.simulation_initial_date is set *and*
self.simulation_time is nonzero (i.e. when running from checkpoint with
load_state()).

stephankramer reviewed (changes requested):
Also added a fix for restarting with simulation_initial_date

jhill1 commented:
I will take over this PR

joewallwork commented:
> I will take over this PR

Thanks Jon! Apologies that I never got back to it.
